org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/IFilterOutput.java

```diff
@@ -41,4 +43,16 @@
 	 */
 	void merge(AbstractInsnNode i1, AbstractInsnNode i2);
 
+	/**
+	 * Marks instruction that should be considered as having branches into the
+	 * given instructions during computation of coverage. The original target
+	 * branches of the original instruction will be ignored.
+	 *
+	 * @param original
+	 *            instruction to be replaced
+	 * @param targets
+	 *            new targets of branches
+	 */
+	void replace(AbstractInsnNode original, List<AbstractInsnNode> targets);
```

Comment:
> The targets have no order and must be unique. Should be a Set.

Actually there is order, which is IMO well defined and very important property of/in `MethodAnalyzer`, otherwise how `merge` works? 😉 In particular `default` of `switch` has number `0`:
https://github.com/jacoco/jacoco/blob/63d55acf9071446e9ea6706d6a65c94d87faca17/org.jacoco.core/src/org/jacoco/core/internal/analysis/MethodAnalyzer.java#L334-L335

And so maybe `List` will be less confusing and more consistent even if doesn't play role in this API ?

----

> The instruction is not replaced, only the CFG edges to its successors is replaced by new edges.

As we discussed earlier and as in implementation - we don't modify CFG, this affects only afterwards computation.

> I know I introduced the term "branch" again.

No problem with term "branch" - it is already in Javadoc.

Right now I realized that with or without renaming of method on `replaceBranches` from this interface alone not clear that replacement of branches changes coverage of instruction and not just branch coverage.

----

> ```
>     /**
>      * Replaces the outgoing branches of a given instruction with new target
> ```

Other methods use term `mark` (consistency?) and state that they affect stage of computations, i.e. not the original CFG and not the propagation of probes, what again as was discussed earlier is quite important in this case. So not sure that this is cleaner/better than current

> ```
> 	/**
> 	 * Marks instruction that should be considered as having branches into the
> 	 * given instructions during computation of coverage.
> ```

Can imagine combination of both:

```java
 	/**
	 * Marks instruction whose outgoing branches should be replaced during
	 * computation of coverage.
 	 * 
 	 * @param source
	 *            instruction which branches should be replaced
 	 * @param newTargets
	 *            new targets of branches
 	 */
	void replaceBranches(AbstractInsnNode source,
```

----

In either case to me seems that important things are missing in this documentation:

* `merge` is performed before `ignore` and `replace`
* `ignore` doesn't affect `merge` and targets in `replace`, but affects source in `replace`
* targets in `replace` are affected by `merge`s
* `replace` overrides `merge`s of `source`